Update sperr

shaomeng opened the pull request:
This PR updates SPERR to the latest release, v0.7.1.

Commit:
update SPERR package

Commit:
update SPERR package

Commit:
remove blank line

Commit:
remove blank line

Commit:
Merge branch 'update-sperr' of github.com:shaomeng/spack into update-sperr

Commit:
Merge branch 'develop' into update-sperr

shaomeng committed:
Merge branch 'develop' into update-sperr

Commit:
Merge branch 'develop' into update-sperr

Commit:
update SPERR to be version 0.7.1

tldahlgren reviewed (approved):
Confirmed the version sha256.

Curious as to why `zstd` variant (non-bundled) is being made and if that change is intentionally applied to all versions?

tldahlgren commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -12,21 +12,20 @@ class Sperr(CMakePackage):
 
     depends_on("git", type="build")
-    depends_on("zstd", type=("build", "link"), when="+zstd")
     depends_on("pkgconfig", type=("build"), when="+zstd")
```
Does the dependency on `pkgconfig` also need to be removed?  If not, then the `when` argument needs to be removed.

shaomeng replied:
It's a good point! I removed the "when" argument!

tldahlgren reviewed (changes requested):
Looks like there are some inconsistencies related to removal of `zstd` variant.

Commit:
a little clean up

shaomeng commented:
> Looks like there are some inconsistencies related to removal of `zstd` variant.

Well, this is a case that may warrant more discussion. SPERR 0.7.1 doesn't' require on `zstd` as a dependency anymore, but previous versions listed here (e.g., v0.6.2) does.  Should I remove reference to previous versions (v0.6.2 etc.) all together? Is there a spack convention handling this situation?

tldahlgren commented:
> > Looks like there are some inconsistencies related to removal of `zstd` variant.
> 
> Well, this is a case that may warrant more discussion. SPERR 0.7.1 doesn't' require on `zstd` as a dependency anymore, but previous versions listed here (e.g., v0.6.2) does. Should I remove reference to previous versions (v0.6.2 etc.) all together? Is there a spack convention handling this situation?

In that case you could make the variant conditional by adding the `when` argument with the appropriate version range (see https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-variants).  Then make the corresponding adjustments to the rest of the lines you've removed.

Even if you deprecate (https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions) the older versions you'd probably still want to make the conditional `zstd` variant changes until those version can be removed.

shaomeng commented:
> In that case you could make the variant conditional by adding the `when` argument with the appropriate version range (see https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-variants). Then make the corresponding adjustments to the rest of the lines you've removed (possibly conditional dependencies https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-dependencies).
> 
> Even if you deprecate (https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions) the older versions you'd probably still want to make the conditional `zstd` variant changes until those version can be removed.

Thanks for the suggestion! I'm almost finishing the conditional variant, but feel puzzled by spack syntax to describe "less than or equal to" a specific version. (I was able to find "great than or equal to" though.)

tldahlgren commented:
> Thanks for the suggestion! I'm almost finishing the conditional variant, but feel puzzled by spack syntax to describe "less than or equal to" a specific version; what's it like? (I was able to find "great than or equal to" though.)

You'll want to use the version range without the lower bound with the meaning `up to and including`, such as:

```
    depends_on("zstd", type=("build", "link"), when="@:0.6.2+zstd")
```

Check out the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges for more information.

robertu94 commented:
We also need to update the version bound for LibPressio-sperr since there were some interface changes in this version. We can update them again once the pr @shaomeng made to update it is merged.

Commit:
bound versions that require zstd

Commit:
minor

shaomeng commented:
> You'll want to use the version range without the lower bound with the meaning `up to and including`, such as:
> 
> ```
>     depends_on("zstd", type=("build", "link"), when="@:0.6.2+zstd")
> ```
> 
> Check out the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges for more information.

Thanks for the pointer! I've updated the PR reflecting the version requirements.

tldahlgren commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -37,11 +40,10 @@ def cmake_args(self):
 
         args = [
             self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
-            self.define_from_variant("USE_ZSTD", "zstd"),
```
Don't you still want this for when `+zstd`?

shaomeng replied:
Yes, this is needed when `+zstd`

shaomeng replied:
Thank you so much for pointing it out! Yes, it is still needed! I've brought it back! 

(Also, the CMake flag `USE_BUNDLED_ZSTD` is actually not needed since version 0.7.x, but leaving it there has no harm. I think it's OK to just leave it there?)

tldahlgren replied:
Depends on how the software builds work but I don't see why not unless keeping it there causes a build failure.  You could also add a variant to enable/disable that feature for versions that use it with the `when` argument, as you've done with `zstd`.

shaomeng replied:
This is a great idea! I've added a variant to toggle it.

tldahlgren reviewed (changes requested):
Don't you still need to use the `zstd` variant for older versions?

robertu94 reviewed (changes requested):
See my previous comment about changing the version upper bound in libpressio-sperr

Commit:
add USE_ZSTD

shaomeng commented:
> See my previous comment about changing the version upper bound in libpressio-sperr

To be clear, your previous comment is about things to do AFTER this PR is merged, is that right?

robertu94 commented:
> > See my previous comment about changing the version upper bound in libpressio-sperr
> 
> To be clear, your previous comment is about things to do AFTER this PR is merged, is that right?

No, we need to change the upper bound NOW (so when we add the new SPERR version we don't break the compilation of LibPressio-SPERR) which the new version does, and again AFTER robertu94/libpressio-sperr#2 is merged to expose the new version.

Commit:
add libpressio-sperr version upbound

shaomeng commented:
> No, we need to change the upper bound NOW (so when we add the new SPERR version we don't break the compilation of LibPressio-SPERR) which the new version does, and again AFTER [robertu94/libpressio-sperr#2](https://github.com/robertu94/libpressio-sperr/pull/2) is merged to expose the new version.

@robertu94 the upper bound specified in [c159075](https://github.com/spack/spack/pull/40626/commits/c159075242567d057fe062f2ee1010b65d4b8a03) should be sufficient, right?

Commit:
minor

Commit:
update libpressio-sperr

shaomeng commented:
> @shaomeng , now that your PR is merged the new version of libpressio-sperr is 0.0.4 with commit 06c1e0555a99963fe45586739c8f06785e17ae50. We can add this version to libpressio-sperr now to save on the number of PRs that need to go to the spack team.

Good idea; I have it added. @robertu94

robertu94 commented on var/spack/repos/builtin/packages/libpressio-sperr/package.py:
```diff
@@ -10,17 +10,19 @@ class LibpressioSperr(CMakePackage):
 
     homepage = "https://github.com/robertu94/libpressio-sperr"
-    url = "https://github.com/robertu94/libpressio-sperr/archive/refs/tags/0.0.1.tar.gz"
+    url = "https://github.com/robertu94/libpressio-sperr/archive/refs/tags/0.0.4.tar.gz"
```
This isn't strictly necessary, but it also doesn't hurt.

tldahlgren reviewed (dismissed):
(Re)confirmed the version sha256.

FYI. I noticed the directives in `sperr` are ordered `version`, `depends_on`, then `variant`s and in `libpressio-sperr` are ordered: `depends_on` then `version`.  The "convention" is to have `version` directives first followed by `variant`s since the latter can depend on the former (https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-variants) then `depends_on` since dependencies can depend on both version(s) and variant(s).

tldahlgren commented:
Looks like there are some style issues that need to be resolved:

```
==> Running black checks
--- var/spack/repos/builtin/packages/sperr/package.py	2023-10-25 20:00:45.020377+00:00
+++ var/spack/repos/builtin/packages/sperr/package.py	2023-10-25 20:00:58.284988+00:00
@@ -41,10 +41,10 @@
         args = [
             self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
             self.define_from_variant("USE_OMP", "openmp"),
             self.define_from_variant("BUILD_CLI_UTILITIES", "utilities"),
             self.define_from_variant("USE_ZSTD", "zstd"),
-            "-DUSE_BUNDLED_ZSTD=OFF", # not in use since version 0.7.x
+            "-DUSE_BUNDLED_ZSTD=OFF",  # not in use since version 0.7.x
             "-DSPERR_PREFER_RPATH=OFF",
             "-DBUILD_UNIT_TESTS=OFF",
         ]
         return args
would reformat var/spack/repos/builtin/packages/sperr/package.py
Oh no! 💥 💔 💥
1 file would be reformatted, 1 file would be left unchanged.
  black found errors
==> Running flake8 checks
var/spack/repos/builtin/packages/sperr/package.py:46: [E261] at least two spaces before inline comment
  flake8 found errors
```

Commit:
address review comments

shaomeng commented:
@tldahlgren This is such good feedback! I've changed the ordering of blocks, improved the format, and also added a variant. Please take a look and let me know if it passes all the tests! Thank you so much!

Commit:
improve format

shaomeng committed:
Merge branch 'develop' into update-sperr

Commit:
remove trailing whitespace

Commit:
reformat

tldahlgren commented on var/spack/repos/builtin/packages/sperr/package.py:
```diff
@@ -12,23 +12,32 @@ class Sperr(CMakePackage):
-    variant("openmp", description="use openmp in 3D inputs", default=True)
+    variant("openmp", description="use OpenMP in 3D inputs", default=True)
     variant("utilities", description="build SPERR CLI utilities", default=True)
+    variant("zstd", description="use ZSTD for more compression", default=True, when="@:0.6.2")
```
So I understand, are you saying `zstd` can now only be enabled for versions up to `0.6.2`?

robertu94 replied:
That is what I understand from @shaomeng, but he should confirm.

shaomeng replied:
Versions up to `0.6.2` can *optionally* take advantage of `zstd` for a little extra compression, which is configurable through CMake. The benefit is however very minimal, so for simplicity, I removed this layer of ZSTD completely in version `0.7.x` and forward, thus there's no `zstd` dependency whatsoever. (There is nothing to be configured w.r.t. `zstd` after version `0.6.2`.)

I of course want the spack recipe to correctly express the underlying logic, and I really appreciate it if you can confirm that the current constraints are doing so (or not). Thanks!

tldahlgren replied:
You can get an idea of how Spack resolves dependencies using `spack spec sperr` plus whatever options you want. 

For example, `spack spec sperr+zstd` for me results in 

```
Input spec
--------------------------------
 -   sperr+zstd

Concretized
--------------------------------
 -   sperr@0.6.2%gcc@12.1.1~bundle_zstd~ipo+openmp+shared+utilities+zstd build_system=cmake build_type=Release generator=make arch=linux-rhel8-x86_64_v3
[+]      ^cmake@3.27.4%gcc@12.1.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-rhel8-x86_64_v3
[+]          ^curl@7.88.1%gcc@12.1.1~gssapi~ldap+libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-rhel8-x86_64_v3
[+]              ^nghttp2@1.52.0%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^ncurses@6.4%gcc@12.1.1~symlinks+termlib abi=none build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^zlib@1.2.13%gcc@12.1.1+optimize+pic+shared build_system=makefile arch=linux-rhel8-x86_64_v3
[+]      ^git@2.41.0%gcc@12.1.1+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^autoconf@2.69%gcc@12.1.1 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-rhel8-x86_64_v3
[+]          ^automake@1.16.5%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^expat@2.5.0%gcc@12.1.1+libbsd build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^libbsd@0.11.7%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]                  ^libmd@1.0.4%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^gettext@0.21.1%gcc@12.1.1+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^bzip2@1.0.8%gcc@12.1.1~debug~pic+shared build_system=generic arch=linux-rhel8-x86_64_v3
[+]              ^libxml2@2.10.3%gcc@12.1.1~python build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^tar@1.34%gcc@12.1.1 build_system=autotools zip=pigz arch=linux-rhel8-x86_64_v3
[+]                  ^pigz@2.7%gcc@12.1.1 build_system=makefile arch=linux-rhel8-x86_64_v3
[+]              ^xz@5.4.1%gcc@12.1.1~pic build_system=autotools libs=shared,static arch=linux-rhel8-x86_64_v3
[+]          ^libiconv@1.17%gcc@12.1.1 build_system=autotools libs=shared,static arch=linux-rhel8-x86_64_v3
[+]          ^libidn2@2.3.4%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^libunistring@1.1%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^libtool@2.4.7%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^m4@1.4.19%gcc@12.1.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-rhel8-x86_64_v3
[+]              ^diffutils@3.9%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^libsigsegv@2.14%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^openssh@9.3p1%gcc@12.1.1+gssapi build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^krb5@1.20.1%gcc@12.1.1+shared build_system=autotools arch=linux-rhel8-x86_64_v3
[+]                  ^bison@3.8.2%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]                  ^findutils@4.9.0%gcc@12.1.1 build_system=autotools patches=440b954 arch=linux-rhel8-x86_64_v3
[+]              ^libedit@3.1-20210216%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]              ^libxcrypt@4.4.35%gcc@12.1.1~obsolete_api build_system=autotools patches=4885da3 arch=linux-rhel8-x86_64_v3
[+]          ^openssl@3.1.1%gcc@12.1.1~docs~shared build_system=generic certs=system arch=linux-rhel8-x86_64_v3
[+]          ^pcre2@10.42%gcc@12.1.1~jit+multibyte build_system=autotools arch=linux-rhel8-x86_64_v3
[+]          ^perl@5.38.0%gcc@12.1.1+cpanm+opcode+open+shared+threads build_system=generic arch=linux-rhel8-x86_64_v3
[+]              ^berkeley-db@18.1.40%gcc@12.1.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel8-x86_64_v3
[+]              ^gdbm@1.23%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]                  ^readline@8.2%gcc@12.1.1 build_system=autotools patches=bbf97f1 arch=linux-rhel8-x86_64_v3
[+]      ^gmake@4.4.1%gcc@12.1.1~guile build_system=autotools arch=linux-rhel8-x86_64_v3
[+]      ^pkgconf@1.9.5%gcc@12.1.1 build_system=autotools arch=linux-rhel8-x86_64_v3
[+]      ^zstd@1.5.5%gcc@12.1.1+programs build_system=makefile compression=none libs=shared,static arch=linux-rhel8-x86_64_v3
```

which shows I've installed the package dependencies already.

And the resulting concretized `sperr` from `spack spec sperr` is just `sperr@0.7.1%gcc@12.1.1~ipo+openmp+shared+utilities build_system=cmake build_type=Release generator=make arch=linux-rhel8-x86_64_v3` (and no `zstd`).

tldahlgren reviewed (approved):
Looking good.  I'm a little confused about the version constraint on `zstd` but will merge if you both think this is ready.

tldahlgren commented:
I went ahead and merged.  Feel free to open another PR if the above results or your testing finds additional changes needed.

shaomeng commented:
> I went ahead and merged. Feel free to open another PR if the above results or your testing finds additional changes needed.

Thanks @tldahlgren , I think it's looking good now!
